src/main/java/org/springframework/integration/kafka/listener/KafkaMessageListenerContainer.java

```diff
@@ -372,89 +385,82 @@ public boolean isLongLived() {
-							Result<KafkaMessageBatch> result = kafkaTemplate.receive(fetchRequests);
-							// process successful messages first
-							Iterable<KafkaMessageBatch> batches = result.getResults().values();
-							for (KafkaMessageBatch batch : batches) {
-								if (!batch.getMessages().isEmpty()) {
-									long highestFetchedOffset = 0;
-									for (KafkaMessage kafkaMessage : batch.getMessages()) {
-										// fetch operations may return entire blocks of compressed messages,
-										// which may have lower offsets than the ones requested
-										// thus a batch may contain messages that have been processed already
-										if (kafkaMessage.getMetadata().getOffset() >= fetchOffsets.get(batch.getPartition())) {
-											messageDispatcher.dispatch(kafkaMessage);
+							if (!listenedPartitions.isEmpty()) {
+								MutableCollection<FetchRequest> fetchRequests = listenedPartitions
+										.collect(new PartitionToFetchRequestFunction());
+								Result<KafkaMessageBatch> result = kafkaTemplate.receive(fetchRequests);
+								// process successful messages first
+								Iterable<KafkaMessageBatch> batches = result.getResults().values();
+								for (KafkaMessageBatch batch : batches) {
+									if (!batch.getMessages().isEmpty()) {
+										long highestFetchedOffset = 0;
+										for (KafkaMessage kafkaMessage : batch.getMessages()) {
+											// fetch operations may return entire blocks of compressed messages,
+											// which may have lower offsets than the ones requested
+											// thus a batch may contain messages that have been processed already
+											if (kafkaMessage.getMetadata().getOffset() >= fetchOffsets
+													.get(batch.getPartition())) {
+												messageDispatcher.dispatch(kafkaMessage);
+											}
+											highestFetchedOffset = Math.max(highestFetchedOffset,
+													kafkaMessage.getMetadata().getNextOffset());
 										}
-										highestFetchedOffset =
-												Math.max(highestFetchedOffset, kafkaMessage.getMetadata().getNextOffset());
-									}
-									fetchOffsets.replace(batch.getPartition(), highestFetchedOffset);
-									// if there are still messages on server, we can go on and retrieve more
-									if (highestFetchedOffset < batch.getHighWatermark()) {
-										partitionsWithRemainingData.add(batch.getPartition());
+										fetchOffsets.replace(batch.getPartition(), highestFetchedOffset);
 									}
 								}
+								// handle errors
+								if (result.getErrors().size() > 0) {
+									// find partitions with leader errors and
+									PartitionIterable<Map.Entry<Partition, Short>> partitionByLeaderErrors = partition(
+											result.getErrors().entrySet(), new IsLeaderErrorPredicate());
+									RichIterable<Partition> partitionsWithLeaderErrors = partitionByLeaderErrors
+											.getSelected().collect(keyFunction);
+									resetLeaders(partitionsWithLeaderErrors);
```

Comment:
I think we should `listenedPartitions.removeAllIterable(partitionsWithLeaderErrors)` before scheduling `resetLeaders`.
We may go to the next iteration here, but the `resetLeaders` won't be evaluated yet, so we are going to visit this error handling block one more time and for the same errors.

Looks like the reset/remove order does not make sense because we are `synchronized (listenedPartitions)` here and in the `addListenedPartitionsIfActive()`.